src/resource/bindless.rs

```diff
@@ -0,0 +1,153 @@
+        vk::DescriptorSetLayoutBinding {
+            binding,
+            descriptor_type: Self::desc_type(),
+            descriptor_count: MAX_BINDLESS_COUNT,
+            stage_flags: vk::ShaderStageFlags::ALL,
+            p_immutable_samplers: std::ptr::null(),
+        }
+    }
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo;
+}
+
+impl Bindlessable for crate::image::Image {
+    fn desc_type() -> vk::DescriptorType {
+        vk::DescriptorType::STORAGE_IMAGE
+    }
+
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo {
+        todo!()
+    }
+}
+
+impl Bindlessable for (crate::image::ImgView, crate::sampler::Sampler) {
+    fn desc_type() -> vk::DescriptorType {
+        vk::DescriptorType::COMBINED_IMAGE_SAMPLER
+    }
+
+    fn descriptor_info(&self) -> vk::DescriptorImageInfo {
+        vk::DescriptorImageInfo {
+            sampler: unsafe { self.1.handle() },
+            image_view: unsafe { self.0.handle() },
+            image_layout: vk::ImageLayout::SHADER_READ_ONLY_OPTIMAL, //TODO don't hardcode
+        }
+    }
+}
+
+pub struct BindlessResource<R: Bindlessable> {
+    key: u32,
+    pool: BindlessPool<R>,
+}
+
+impl<R: Bindlessable> Drop for BindlessResource<R> {
+    fn drop(&mut self) {
+        self.pool.with(|p| p.take(self.key));
+    }
+}
+
+impl<R: Bindlessable> BindlessResource<R> {
+    pub fn index(&self) -> u32 {
+        self.key
+    }
+}
+
+struct BindlessPoolInner<R> {
+    items: Vec<Option<R>>,
+    free: Vec<u32>,
+    descriptor_set: crate::DescriptorSet,
+}
+
+impl<R: Bindlessable> BindlessPoolInner<R> {
+    fn update_ds(&mut self, i: u32, r: &R) {
+        let vk_writes = vk::WriteDescriptorSet {
+                s_type: vk::StructureType::WRITE_DESCRIPTOR_SET,
+                p_next: std::ptr::null(),
+                dst_set: self.descriptor_set.handle,
+                dst_binding: 0,
+                dst_array_element: i,
+                descriptor_count: 1,
+                descriptor_type: R::desc_type(),
+                p_image_info: &r.descriptor_info() as *const _,
+                p_buffer_info: std::ptr::null(),
+                p_texel_buffer_view: std::ptr::null(),
+            };
+
+        unsafe {
+            self.descriptor_set.device.update_descriptor_sets(&[vk_writes], &[]);
+        }
+    }
+
+
+    fn take(&mut self, key: u32) -> Option<R> {
+        if key <= self.items.len() as _ {
+            None
+        } else {
+            self.items[key as usize].take().and_then(|ob| {
+                self.free.push(key);
+                Some(ob)
+            })
+        }
+    }
+}
+
+pub struct BindlessPool<R> {
+    inner: Arc<Mutex<BindlessPoolInner<R>>>,
+}
+
+impl<P: Bindlessable> BindlessPool<P> {
+    fn with<F: FnOnce(&mut BindlessPoolInner<P>) -> R, R>(&self, f: F) -> R {
+        let mut inner = self.inner.lock().unwrap();
+        f(&mut inner)
+    }
+
+    pub fn alloc(&mut self, item: P) -> BindlessResource<P> {
```

Comment:
I think it's reasonable to have a method in `Bindlessable` like the following

```rs
pub trait Bindlessable {
    fn into_bindless(self, pool: &BindlessPool<Self>) -> BindlessResource<Self>
}
```

that just calls this method. Both can probably exist just fine.